upload/views/combined_upload.py

```diff
@@ -0,0 +1,171 @@
+            code=request.data.get("code"),
+        )
+        commit_report_serializer = CommitReportSerializer(data=commit_report_data)
+        if not commit_report_serializer.is_valid():
+            return Response(
+                commit_report_serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
+
+        inc_counter(
+            API_UPLOAD_COUNTER,
+            labels=generate_upload_prometheus_metrics_labels(
+                action="coverage",
+                endpoint="combined_upload",
+                request=self.request,
+                is_shelter_request=self.is_shelter_request(),
+                position="create_report",
+            ),
+        )
+        report = self.create_report(commit_report_serializer, repository, commit)
+
+        # Do upload
+        upload_data = dict(
+            ci_url=request.data.get("build_url"),
+            env=request.data.get("env_vars"),
+            flags=request.data.get("flags"),
+            ci_service=request.data.get("ci_service"),
+            job_code=request.data.get("job_code"),
+            name=request.data.get("name"),
+        )
+
+        upload_serializer = UploadSerializer(data=upload_data)
+        if not upload_serializer.is_valid():
+            return Response(
+                upload_serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
+
+        inc_counter(
+            API_UPLOAD_COUNTER,
+            labels=generate_upload_prometheus_metrics_labels(
+                action="coverage",
+                endpoint="combined_upload",
+                request=self.request,
+                is_shelter_request=self.is_shelter_request(),
+                position="create_upload",
+            ),
+        )
+        upload = self.create_upload(upload_serializer, repository, commit, report)
+
+        inc_counter(
+            API_UPLOAD_COUNTER,
+            labels=generate_upload_prometheus_metrics_labels(
+                action="coverage",
+                endpoint="combined_upload",
+                request=self.request,
+                is_shelter_request=self.is_shelter_request(),
+                position="end",
+            ),
+        )
+
+        if upload:
+            commitid = upload.report.commit.commitid
+            upload_repository = upload.report.commit.repository
+            url = f"{settings.CODECOV_DASHBOARD_URL}/{upload_repository.author.service}/{upload_repository.author.username}/{upload_repository.name}/commit/{commitid}"
+            archive_service = ArchiveService(upload_repository)
+            raw_upload_location = archive_service.create_presigned_put(
+                upload.storage_path
+            )
+            return Response(
+                {
+                    "url": url,
+                    "raw_upload_location": raw_upload_location,
+                },
+                status=status.HTTP_201_CREATED,
+            )
+        else:
+            return Response(
```

Comment:
I would early-return using `if not upload`, similar to other places.

Reply:
Done!